torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+
+    builder.setInsertionPointAfter(op);
+    llvm::SmallVector<mlir::NamedAttribute> call_attrs{
+        {
+            builder.getStringAttr("call_target_name"),
+            builder.getStringAttr("stablehlo.composite"),
+        },
+        {
+            builder.getStringAttr("called_computations"),
+            builder.getArrayAttr(mlir::FlatSymbolRefAttr::get(
+                builder.getContext(), impl_func.getSymName())),
+        },
+        {
+            builder.getStringAttr("composite.backend_config"),
+            builder.getDictionaryAttr(llvm::SmallVector<mlir::NamedAttribute>{
+                {
+                    builder.getStringAttr("attributes"),
+                    BuildDictionaryAttrFromJsonMap(builder,
+                                                   output_metadata.attrs),
+                },
+                {
+                    builder.getStringAttr("name"),
+                    builder.getStringAttr(output_metadata.name),
+                },
+            }),
+        },
+    };
+    // Inserts composite call op.
+    mlir::Operation* composite_call_op =
+        builder.create<mlir::stablehlo::CustomCallOp>(
+            op->getLoc(), impl_func.getFunctionType().getResults(), args,
+            call_attrs);
+
+    // Updates all users of this op's result(s) to use the results(s) of impl
+    // func call.
+    for (size_t i = 0; i < op->getNumResults(); ++i) {
+      mlir::OpResult result = op->getResult(i);
+      mlir::OpResult new_result = composite_call_op->getResult(i);
+      for (mlir::OpOperand& use : result.getUses()) {
+        use.getOwner()->setOperand(use.getOperandNumber(), new_result);
+      }
+    }
+
+    // The unused scope_ops will be eliminated with canonicalizer.
+  }
+};
+
+class RemoveXlaMarkTensorOpsPass
+    : public mlir::OperationPass<mlir::func::FuncOp> {
+ public:
+  explicit RemoveXlaMarkTensorOpsPass()
+      : mlir::OperationPass<mlir::func::FuncOp>::OperationPass(
+            mlir::TypeID::get<RemoveXlaMarkTensorOpsPass>()) {}
+
+  ~RemoveXlaMarkTensorOpsPass() override = default;
+
+  void runOnOperation() override {
+    mlir::func::FuncOp func_op = getOperation();
+    llvm::SmallVector<mlir::Operation*> ops_to_erase;
+
+    for (auto mark_tensor_op :
+         func_op.getOps<mlir::stablehlo::CustomCallOp>()) {
+      if (!IsXlaMarkTensorOp(mark_tensor_op.getOperation())) {
+        continue;
+      }
+      mlir::Value original_value = mark_tensor_op.getOperand(0);
+
+      llvm::SmallVector<std::tuple<mlir::Operation*, size_t>> uses;
+      for (mlir::OpOperand& use : mark_tensor_op.getResult(0).getUses()) {
+        uses.push_back({use.getOwner(), use.getOperandNumber()});
+      }
+
+      for (auto [use_op, operand_number] : uses) {
+        use_op->setOperand(operand_number, original_value);
+      }
+      ops_to_erase.push_back(mark_tensor_op.getOperation());
+    }
+    for (auto* op : ops_to_erase) {
+      op->erase();
+    }
```

Comment:
Can we add code comments? I'm not positive what this is doing.. Is this just replacing the uses of the custom call result with its operand? If so something like this should manage the intermediate logic:

```
for (auto op : func_op.getOps<mlir::stablehlo::CustomCallOp>()) {
  if (!IsXlaMarkTensorOp(op.getOperation())) continue;
  for (auto i = 0; i < op.getNumResults(); ++i) {
    Value result = op.getResult(i);
    result.replaceAllUsesWith(op.getOperand(i));
  }
}
```

Note this is coded for the pairwise input-output if we end up going with a design that permits specifying multiple inputs to an abstraction at once.

Reply:
Also instead of explicitly calling erase we can let DCE during canonicalization take care of that, or just call erase inline - we likely don't need to keep a vector of deleted nodes just to delete later?

Looks like we already have a canonicalize pass that runs with this below.

Reply:
Yes you're right. I'm not aware of the `replaceAllUsesWith` APIs.